### ISSUE TITLE
c++ H5File Constructor Error

### ISSUE THREAD

**Issue by lylong94** [HDFGroup/hdf5#5191]
When create H5File object use constructor [`H5File(const H5std_string &name, unsigned int flags, const FileAccPropList &access_plist);`](https://github.com/HDFGroup/hdf5/blob/f2642985d8c23ff7e876c6228c7cc0cf20515923/c%2B%2B/src/H5File.h#L34) may pass wrong `const char*` string to `H5Fcreate`.

The STL String default without the NULL character `\0`. So in the row [`p_get_file(name.c_str(), flags, create_plist, access_plist);`](https://github.com/HDFGroup/hdf5/blob/f2642985d8c23ff7e876c6228c7cc0cf20515923/c%2B%2B/src/H5File.cpp#L114), simply pass it by `name.c_str()` may not correct and add some characters.

The simplest way is append `\0` before call `c_str()`.

**Comment by lylong94** [comment]
It appears not only in `H5Fil`, but also in all `string.c_str()`.

**Comment by bmribler** [comment]
@lylong94 The c_str() method is guaranteed by the C++ standard to return a properly null-terminated string, so no manual null terminator is needed. If you're experiencing issues with file creation, could you provide more details about the error you're seeing?

**Comment by lylong94** [comment]
@brtnfld You are right. I checked the standard and the c_str() method is guaranteed return null-terminated.
The project is a Qt5 project, compiled using MSVC v143 in Qt Creator. In Release mode, I found that the program failed to find the HDF5 file and throw an error when reading the file. Upon inspection, I discovered that the filename had an unexpected suffix. However, the file path printed before being passed to the function was correct. Adding a null terminator ('\0') to the `std::string` resolved the issue.
I'm still quite confused, but I've left that team so I can't demonstrate the error. Based on standard practices, this doesn't seem to be an error in HDF5.It may remain an unsolved mystery. ╮(╯_╰)╭

**Comment by bmribler** [comment]
@lylong94 Thanks for responding, even after you've left the team.  I'm curious about the behavior though.  I might investigate it a bit when I have some time.  Would you be able to share the file?  If not, maybe just the name of the file or the unexpected suffix you mentioned so I can imitate similar file.
